src/datasource.ts

```diff
@@ -264,25 +321,156 @@ export default class CogniteDatasource {
+      for (let option in options) {
+        urlEnd += "&" + option + "=" + options[option];
+      }
     }
+
     return this.backendSrv.datasourceRequest({
       url: this.url + urlEnd,
       method: "GET",
     }).then((result: { data: TimeSeriesResponse }) =>
       result.data.data.items.map(timeSeriesResponseItem => (
         {
           text: timeSeriesResponseItem.name,
-          value: timeSeriesResponseItem.name
+          value: (type==="Asset") ? '' + timeSeriesResponseItem.id : timeSeriesResponseItem.name
+        }))
+      );
+  }
+
+  findAssetTimeseries(target, panelCtrl) {
+    this.backendSrv.datasourceRequest({
+      url: this.url + `/cogniteapi/${this.project}/timeseries/search?` +
+      `${(target.assetQuery.includeSubtrees) ? "assetSubtrees" : "assetIds"}=[${target.assetQuery.target}]&` +
+      `isString=false&` +
+      `limit=1000`,
+      method: "GET",
+    }).then((result: { data: TimeSeriesResponse }) => {
+      target.assetQuery.timeseries = result.data.data.items.map(ts => {
+        ts.selected = true;
+        return ts;
+      });
+    }).then(() => panelCtrl.refresh());
+  }
+
+  filterOnAssetTimeseries(target) {
+    const filterOptions = this.parse(target.expr);
+
+    for (let ts of target.assetQuery.timeseries) {
+      ts.selected = true;
+      for (let filter of filterOptions.filters) {
+        if (filter.type === "=~") {
+          const val = _.get(ts,filter.property);
+          const regex = "^" + filter.value + "$";
+          if (val === undefined || !val.match(regex)) {
+            ts.selected = false;
+            break;
+          }
+        } else if (filter.type === "!~") {
+          const val = _.get(ts,filter.property);
+          const regex = "^" + filter.value + "$";
+          if (val === undefined || val.match(regex)) {
+            ts.selected = false;
+            break;
+          }
+        } else if (filter.type === "!=") {
+          const val = _.get(ts,filter.property);
+          if (val === undefined || (val == filter.value)) {
+            ts.selected = false;
+            break;
+          }
+        } else if (filter.type === "=") {
+          const val = _.get(ts,filter.property);
+          if (val === undefined || (val != filter.value)) {
+            ts.selected = false;
+            break;
+          }
+        }
+      }
+    }
+
+    if (filterOptions.aggregation) {
+      target.aggregation = filterOptions.aggregation;
+    }
+    if (filterOptions.granularity) {
+      target.granularity = filterOptions.granularity;
+    }
+  }
+
+  parse(customQuery) {
+    // replace variables with their values
+    for (let templateVariable of this.templateSrv.variables) {
+      customQuery = customQuery.replace("[[" + templateVariable.name + "]]", templateVariable.current.value);
+    }
+
+    let filtersOptions = {
+      filters: [],
+      granularity: '',
+      aggregation: ''
+    };
+
+    const timeseriesRegex = /^timeseries\{(.*)\}(?:\[(.*)\])$/;
```

Comment:
I don't have a good alternative solution for the time being, but this section looks like a nightmare to maintain 👿

Reply:
Added a comment to describe what this regex is doing. 
I think for the time being this is a fine way to approach this, but if we want to add more complexity to the querying, then I agree that we will have to change some things.